src/client/api/metrics.c

```diff
@@ -0,0 +1,146 @@
+/*
+ * (C) Copyright 2020-2023 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/utsname.h>
+#include <daos/common.h>
+#include <daos/job.h>
+#include <daos/tls.h>
+#include <gurt/telemetry_common.h>
+#include <gurt/telemetry_consumer.h>
+#include <gurt/telemetry_producer.h>
+
+#define INIT_JOB_NUM		1024
+bool daos_client_metric;
+bool daos_client_metric_retain;
+
+#define MAX_IDS_SIZE(num) (num * D_TM_METRIC_SIZE)
+/* The client side metrics structure looks like
+ * root/job_id/pid/....
+ */
+int
+dc_tm_init(void)
+{
+	int    metrics_tag;
+	pid_t  pid;
+	int    rc;
+
+	d_getenv_bool(DAOS_CLIENT_METRICS_ENV, &daos_client_metric);
+	if (!daos_client_metric)
+		return 0;
+
+	rc = dc_tls_key_create();
+	if (rc)
+		D_GOTO(out, rc);
+
+	metrics_tag = D_TM_OPEN_OR_CREATE;
+	d_getenv_bool(DAOS_CLIENT_METRICS_RETAIN_ENV, &daos_client_metric_retain);
+	if (daos_client_metric_retain)
+		metrics_tag |= D_TM_RETAIN_SHMEM;
+	else
+		metrics_tag |= D_TM_RETAIN_SHMEM_IF_NON_EMPTY;
+
+	rc = d_tm_init(DC_TM_JOB_ROOT_ID, MAX_IDS_SIZE(INIT_JOB_NUM), metrics_tag);
+	if (rc != 0) {
+		DL_ERROR(rc, "init job root id.");
+		return rc;
+	}
+
+	pid = getpid();
+	D_INFO("INIT %s/%u metrics\n", dc_jobid, pid);
+
+	/** create new shmem space for per-pool metrics */
+	rc = d_tm_add_ephemeral_dir(NULL, MAX_IDS_SIZE(INIT_JOB_NUM), "%s/%u",
```

Comment:
Was just chatting with @kjacque about this... I think we need the per-job ephemeral dir, but we also need per-pid ephemeral dirs under it. This would allow the agent to prune old per-pid telemetry without disturbing other processes that are using the same jobid. I am working on that part of it now.

I think we also need to adjust the size of these segments, something along these lines:
  - client root: D_TM_METRIC_SIZE * NUM_SUPPORTED_JOBIDS (256 ?)
  - per-jobid: D_TM_METRIC_SIZE * NUM_SUPPORTED_JOB_PROCS (1024 ?)
  - per-proc: D_TM_METRIC_SIZE * NUM_CLIENT_METRICS (not sure how to calc this)
  
  We may want to make those configurable someday, but I don't think we need to worry about it now.

Reply:
I think what we have now is fine to start. I can play around with any changes to make pruning work better in my follow-on patch.

Reply:
Trying to add per pool metrics on client side.

Reply:
@mjmac  Just added per pool metrics for fetch/update bytes, please let me know if you need further improvements.

Reply:
@wangdi1 I played around with this last night... I think there is a bug in the per-pool metrics initialization if a label is used. I had to move the call to dc_pool_metrics_start() until after the label is resolved into a UUID, otherwise the pool UUID is always the nil UUID (0000). Not sure if this is the best place to put it, but it works in my testing:

```
@@ -651,10 +653,6 @@ init_pool(const char *label, uuid_t uuid, uint64_t capas, const char *grp,
        if (rc != 0)
                D_GOTO(err_pool, rc);

-       rc = dc_pool_metrics_start(pool);
-       if (rc != 0)
-               D_GOTO(err_pool, rc);
-
        /** Agent configuration data from pool->dp_sys->sy_info */
        /** sy_info.provider */
        /** sy_info.interface */
@@ -699,6 +697,11 @@ dc_pool_connect_internal(tse_task_t *task, daos_pool_info_t *info, const char *l
	rc = dc_pool_choose_svc_rank(label, pool->dp_pool, &pool->dp_client, &pool->dp_client_lock,
				     pool->dp_sys, &ep);
	if (rc != 0) {
		DL_CDEBUG(rc == -DER_NONEXIST, DB_PL, DLOG_ERR, rc,
			  DF_UUID ":%s: cannot find pool service", DP_UUID(pool->dp_pool),
			  label ? label : "");
                goto out;
        }

+       /** Start the pool metrics now that we are guaranteed to have a UUID */
+       rc = dc_pool_metrics_start(pool);
+       if (rc != 0)
+               D_GOTO(out, rc);
+
```

I also was hitting a segfault on dc_tm_fini() when metrics aren't retained and the shm is cleaned up. I made the following change and it fixed the segfault:
```
--- a/src/gurt/telemetry.c
+++ b/src/gurt/telemetry.c
@@ -2405,14 +2405,18 @@ failure:
 }

 static void
-invalidate_link_node(struct d_tm_node_t *node)
+invalidate_link_node(struct d_tm_shmem_hdr *parent, struct d_tm_node_t *node)
 {
+       struct d_tm_metric_t *link_metric;
+
        if (node == NULL || node->dtn_type != D_TM_LINK)
                return;

        node->dtn_name = NULL;
-       if (node->dtn_metric != NULL)
-               node->dtn_metric->dtm_data.value = 0;
+       if (node->dtn_metric != NULL) {
+               link_metric = conv_ptr(parent, node->dtn_metric);
+               link_metric->dtm_data.value = 0;
+       }
```

I don't fully understand all of the conv_ptr() changes... It seems very tricky to remember which pointers can be accessed directly and which need to be converted for the shm offset, I guess.

Reply:
Thanks! @mjmac  Sorry, I did not try label case. Yes, you fix seems good to me. 

Hmm, it seems some share memory is not being freed if the metrics are not retained. let me update the patch. Thanks